addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js

```diff
@@ -2891,27 +2908,62 @@ export class OdooEditor extends EventTarget {
      */
     _onPaste(ev) {
         ev.preventDefault();
+
+        const sel = this.document.getSelection();
+        const cleanHtmlOptions = {
+            selectionStartInP: !!closestElement(sel.anchorNode, 'p'),
+            selectionEndInP: !!closestElement(sel.focusNode, 'p')
+        };
+
+        if (!sel.isCollapsed) {
+            this.deleteRange(sel);
+            this.historyStep();
+        }
+
         const files = getImageFiles(ev.clipboardData);
-        const clipboardHtml = ev.clipboardData.getData('text/html');
+        const htmlClipboardData = ev.clipboardData.getData('text/html');
+        const plainTextData = ev.clipboardData.getData('text/plain');
+        let cleanHtmlData = null;
+        let shouldForceDeleteFoward = false;
+
+        if (htmlClipboardData) {
+            cleanHtmlData = this._cleanHtmlClipboardData(htmlClipboardData, cleanHtmlOptions);
+
+            const paragraphRegex = /^<p>(.+)<\/p>$/;
+            // Avoid inserting the opening and closing paragraph tags
+            // in order to ensure we don't add unwanted line breaks
+            if (cleanHtmlData.match(/^<p>/)) {
+                shouldForceDeleteFoward = true;
+            }
+        }
+
         if (files.length) {
             this.addImagesFiles(files);
-        } else if (clipboardHtml) {
-            this.execCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
+        } else if (cleanHtmlData && plainTextData !== cleanHtmlData) {
+            this.historyPauseSteps();
+            this.execCommand('insertHTML', cleanHtmlData, shouldForceDeleteFoward);
+            this.historyUnpauseSteps();
+            this.historyStep();
         } else {
-            const text = ev.clipboardData.getData('text/plain');
-            const splitAroundUrl = text.split(URL_REGEX);
+            const splitAroundUrl = plainTextData.split(URL_REGEX);
             const linkAttributes = this.options.defaultLinkAttributes || {};
+            const selectionIsInsideALink = !!closestElement(sel.anchorNode, 'a');
```

Comment:
This will only be true if the selection _starts_ in a link. What if it doesn't start in a link but ends in one?